### ISSUE TITLE
Memory Leak IThreadPoolWorkItem[]

### ISSUE THREAD

**Issue by HeikoWeiss** [mono/mono#9631]
## Steps to Reproduce

1. Run a thread in a windows forms app main form.
2. cyclic invoke the main form from thread to open a new form (with some labels)
3. wait in thread till the new form is closed
4. close the new form with a timer (e.g. after one second).
5. go to step 2

I do not know, where the memory leak is but,

if I run mono about 1 hour with profiling enabled (--gc=sgen --profile=log:heapshot) I see with mprof-report that the count and the bytes used of IThreadPoolWorkItem[] increases slowly.
Thats the only object which increaes count and memory.

### Current Behavior

Memory leak.

## On which platforms did you notice this

[ ] macOS
[X] Linux
[ ] Windows

**Version Used**:

mono 5.11.0.345

**Comment by bpiepiora** [comment]
Hello Heiko,

I have the same issue when using the TPL with `Parallel.Invoke()` or `Task.Start()` under Linux. Are there any ideas on that topic or do you have a workaround for that, except restarting the app of course ;-) ?

Even the latest mono version 5.18.0.240 has this issue.

**Comment by bpiepiora** [comment]
@marek-safar Maybe the label should be changed or should I create another issue?

**Comment by HeikoWeiss** [comment]
Sorry Marek,

I do not have a solution or workaround about that.

Maybe you should ceate another issue.

**Comment by marek-safar** [comment]
@bpiepiora if you can reproduce the issue outside of Windows Forms app, please create a new issue, ideally with self contained repro

**Comment by ztzg** [comment]
We did observe the same symptoms on a project which does not involve Windows Forms.  Here is an analysis and "hacky" workaround:

https://github.com/ztzg/mono/commit/61780a2d26e2

(Switching from 5.16 to 5.18 "fixed" the issue for us, but I believe it was purely incidental and could show up in other circumstances.)

Also linking to https://stackoverflow.com/q/51114767, "Debugging a memory leak in a long-running mono/C# application," which seems related as it mentions a "massive amount of `IThreadPoolWorkItems`."

**Comment by bpiepiora** [comment]
Thanks @ztzg for the workaround! Since we are deploying this in a fixed environment, the patch can definitely help us.

Btw: I tried 5.18.0.225 and 5.18.0.240 and both versions have the memory leak. Also the problem only seems to apply to ARM machines, which is following your commit description as far as I understand.

So how can we go on from here? I can't decide how "hacky" your workaround truly is, but I think this issue should be addressed somehow.

**Comment by ztzg** [comment]
Hi Benjamin,

> Thanks @ztzg for the workaround! Since we are deploying this in a
> fixed environment, the patch can definitely help us.

Not sure about the patch, but at least you have a credible diagnostic
for what you are observing.

> Btw: I tried 5.18.0.225 and 5.18.0.240 and both versions have the
> memory leak. Also the problem only seems to apply to ARM machines,
> which is following your commit description as far as I understand.

The root cause is that SGen still scans stacks conservatively—
independently of the platform.  But yes, the ARM backend having such a
large initialized "hole" in its LMF certainly doesn't help.

The fact that it still happens on 5.18 in your case probably means that
you are hitting it via some other path.

> So how can we go on from here? I can't decide how "hacky" your
> workaround truly is, but I think this issue should be addressed
> somehow.

Pretty hacky!  As in: zero chances of being merged into mainline :)
But it should be safe.

We know that in our case, the offending LMF was always coming from a
call to Monitor.Wait, so special-casing that method in the JIT was
enough.  (We did not dig any further, as 5.18 made the issue go away.)

As mentioned above, your software may be triggering the leak via another
method—in which case the patch won't help.

(Saying which method(s) would require some "hands-on" analysis.)

> Another idea: Have you tried the Boehm GC on that issue?

Nope, and we want to stay with a generational collector.

You should also note that Boehm is the ultimate conservative GC, so that
would be moving in the wrong direction :)

(I won't tell you not to try, though, as you may be lucky.  But the data
structure in question is quite nasty wrt. conservative GCs.)

Cheers, -D

**Comment by bpiepiora** [comment]
Okay, thanks for the detailled answer. I am trying to isolate the part in my application, that causes the memory to grow.

As far as I understand now, the right solution would be to rework the GC. Is it realistic that this will be implemented in the future at all? I have no idea how much effort this would be.

**Comment by ztzg** [comment]
> Okay, thanks for the detailled answer. I am trying to isolate the part
> in my application, that causes the memory to grow.

Right; doing so is unfortunately far from trivial.

> As far as I understand now, the right solution would be to rework the
> GC. Is it realistic that this will be implemented in the future at
> all? I have no idea how much effort this would be.

The initial plan was for SGen to support precise stack scanning.  From
sgen-mono.c:

    /* If set, mark stacks conservatively, even if precise marking is possible */
    static gboolean conservative_stack_mark = FALSE;

OTOH, it has been stuck like this since 2012:

    ///* Keep this the default for now */
    /* Precise marking is broken on all supported targets. Disable until fixed. */
    conservative_stack_mark = TRUE;

Perhaps @marek-safar can tell us?

Best, -D

**Comment by marek-safar** [comment]
That's correct, Mono GC scan stack conservatively and it won't change anytime soon
